### PR TITLE
[pulsar-broker] namespace resources use metadata-store api

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation
This PR is on top of #9338, with this change, namespace-resources will start using metadata-api and remove zk-dependency.

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java [hunk 1]
```diff
@@ -856,172 +764,52 @@ protected void internalSetSubscriptionExpirationTime(int expirationTime) {
...
         validateAdminAccessForTenant(namespaceName.getTenant());
         validatePoliciesReadOnlyAccess();
-
-        if (!AutoTopicCreationOverride.isValidOverride(autoTopicCreationOverride)) {
-            throw new RestException(Status.PRECONDITION_FAILED, "Invalid configuration for autoTopicCreationOverride");
-        }
-        if (maxPartitions > 0 && autoTopicCreationOverride.defaultNumPartitions > maxPartitions) {
-            throw new RestException(Status.NOT_ACCEPTABLE,
-                    "Number of partitions should be less than or equal to " + maxPartitions);
+        if (autoTopicCreationOverride != null) {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:778]: this `if` statement was not present before, can it be `null` here ?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/v1/V1_AdminApiTest.java [hunk 2]
```diff
@@ -2041,4 +2044,9 @@ public void testCompactionStatus() throws Exception {
         assertTrue(admin.topics().compactionStatus(topicName)
             .lastError.contains("Failed at something"));
     }
+
+    private void clearCache() {
+        // (MetadataCacheImpl<ClusterData>) pulsar.getPulsarResources().
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/v1/V1_AdminApiTest.java:2049]: nit: remove comment

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/cache/impl/MetadataCacheImpl.java [hunk 3]
```diff
@@ -253,4 +255,23 @@ public void accept(Notification t) {
...
         }
     }
+
+    private CompletableFuture<Void> executeWithRetry(Supplier<CompletableFuture<Void>> op, String key) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        op.get().thenAccept(r -> result.complete(null)).exceptionally((ex) -> {
+            if (ex.getCause() instanceof BadVersionException) {
+                // if resource is updated by other than metadata-cache then metadata-cache will get bad-version
+                // exception. so, try to invalidate the cache and try one more time.
+                objCache.synchronous().invalidate(key);
```
Comment (eolivelli) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/cache/impl/MetadataCacheImpl.java:265]: can we call here this `synchronous` operation ?
do we risk a deadlock ?

**Review by eolivelli — COMMENTED** [review]
Great work.
Code really looks cleaner

I left a couple of questions

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java (hunk 1 above)
Comment (rdhabalia) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:778]: yes, I merged two methods which had duplicate code:
`internalRemoveAutoTopicCreation` and `internalSetAutoSubscriptionCreation`

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/cache/impl/MetadataCacheImpl.java (hunk 3 above)
Comment (rdhabalia) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/cache/impl/MetadataCacheImpl.java:265]: no, it just cleans up the cache synchronously.

**Review by eolivelli — APPROVED** [review]
Lgtm

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/cache/impl/MetadataCacheImpl.java [hunk 4]
```diff
@@ -118,7 +120,7 @@ private MetadataCacheImpl(MetadataStore store, MetadataSerde<T> serde) {
 
     @Override
     public CompletableFuture<Void> readModifyUpdateOrCreate(String path, Function<Optional<T>, T> modifyFunction) {
-        return objCache.get(path)
+        return executeWithRetry(() -> objCache.get(path)
```
Comment (merlimat) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/cache/impl/MetadataCacheImpl.java:123]: This part is already in master, can you merge or rebase?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java [hunk 5]
```diff
@@ -418,7 +382,7 @@ protected Policies getNamespacePolicies(NamespaceName namespaceName) {
         try {
             final String namespace = namespaceName.toString();
             final String policyPath = AdminResource.path(POLICIES, namespace);
-            Policies policies = policiesCache().get(policyPath)
+            Policies policies = namespaceResources().get(policyPath)
```
Comment (merlimat) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:385]: nit: ideally namespace resources should take a namespace name and internally resolve the path on metadata store.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java [hunk 6]
```diff
@@ -132,20 +131,18 @@ protected void internalCreateNamespace(Policies policies) {
...
             }
-            policiesCache().invalidate(path(POLICIES, namespaceName.toString()));
-
-            zkCreateOptimistic(path(POLICIES, namespaceName.toString()), jsonMapper().writeValueAsBytes(policies));
+            namespaceResources().create(path(POLICIES, namespaceName.toString()), policies);
             log.info("[{}] Created namespace {}", clientAppId(), namespaceName);
-        } catch (KeeperException.NodeExistsException e) {
-            log.warn("[{}] Failed to create namespace {} - already exists", clientAppId(), namespaceName);
+        } catch (NotFoundException e) {
+            log.error("[{}] namespace already exists {}", clientAppId(), namespaceName, e);
```
Comment (merlimat) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:145]: The exception and the log message seem to have diverged now

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java [hunk 7]
```diff
@@ -1021,4 +1026,75 @@ protected void validateBrokerName(String broker) throws MalformedURLException {
...
         }
     }
+
+    /*
+     * Get the list of namespaces (on every cluster) for a given property.
+     *
+     * @param property the property name
+     * @return the list of namespaces
+     */
+    protected List<String> getListOfNamespaces(String property) throws Exception {
```
Comment (merlimat) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java:1036]:
```suggestion
    protected List<String> getListOfNamespaces(String tenant) throws Exception {
```

**Review by sijie — APPROVED** [review]

**Pull request merged**